### PR TITLE
kernel: arch: introduce k_float_enable()

### PR DESCRIPTION
**Pull request by katsuster**
This patch introduce new API to enable FPU of thread. This is pair of
existed k_float_disable() API. And also add empty arch_float_enable()
into each architectures that have arch_float_disable(). The arc and
riscv already implemented arch_float_enable() so I do not touch
these implementations.

Motivation: Current Zephyr implementation does not allow to use FPU
on main and other system threads like as work queue. Users need to
create an other thread with K_FP_REGS for floating point programs.
Users can use FPU more easily if they can enable FPU on running
threads.

### CODE REVIEW

**Comment by katsuster** [comment]
checkpatch got angry if I used ENOSYS in arch_float_enalbe() stub functions...

File: arch/arc/core/thread.c [hunk 1]
```diff
@@ -231,7 +231,6 @@ int arch_float_disable(struct k_thread *thread)
 	return 0;
 }
 
-
```
Comment (ioannisg) [inline arch/arc/core/thread.c:234]: you could discard this white space diff :)

File: arch/arm/core/aarch32/thread.c [hunk 2]
```diff
@@ -353,6 +353,11 @@ int arch_float_disable(struct k_thread *thread)
 
 	return 0;
 }
+
+int arch_float_enable(struct k_thread *thread)
+{
+	return -ENOSYS;
```
Comment (ioannisg) [inline arch/arm/core/aarch32/thread.c:359]: That's fine for me. Please, add an inline comment that "this is not supported in Cortex-M" or something similar

Comment (ioannisg) [reply arch/arm/core/aarch32/thread.c:359]: FYI: Cortex-R does not support FPU in general, at the moment.

File: arch/riscv/core/thread.c [hunk 3]
```diff
@@ -151,7 +151,6 @@ int arch_float_disable(struct k_thread *thread)
 	return 0;
 }
 
-
```
Comment (ioannisg) [inline arch/riscv/core/thread.c:154]: same here, pls, revert this white space diff

File: kernel/include/kernel_arch_interface.h [hunk 4]
```diff
@@ -180,6 +180,17 @@ void arch_switch_to_main_thread(struct k_thread *main_thread, char *stack_ptr,
  * @retval -EINVAL If the floating point disabling could not be performed.
  */
 int arch_float_disable(struct k_thread *thread);
+
+/**
+ * @brief Enable floating point context preservation
+ *
+ * The function is used to enable the preservation of floating
+ * point context information for a particular thread.
+ *
```
Comment (ioannisg) [inline kernel/include/kernel_arch_interface.h:189]: Add some text, also, here, that this API is optional , and depends on the architecture.

**Review by ioannisg — COMMENTED** [review]
Seems to be in the right direction.

File: arch/arc/core/thread.c (hunk 1 above)
Comment (katsuster) [reply arch/arc/core/thread.c]: Oops... fixed.

File: arch/riscv/core/thread.c (hunk 3 above)
Comment (katsuster) [reply arch/riscv/core/thread.c]: Thanks. fixed.

File: arch/arm/core/aarch32/thread.c (hunk 2 above)
Comment (katsuster) [reply arch/arm/core/aarch32/thread.c:360]: OK, added such information.

File: kernel/include/kernel_arch_interface.h (hunk 4 above)
Comment (katsuster) [reply kernel/include/kernel_arch_interface.h:191]: I added such text.

**Comment by katsuster** [comment]
Conflicting k_float_enable() on x86 platform... I will check.

**Comment by katsuster** [comment]
Changes in v2.
  - x86/ia32: change k_float_enable() -> z_float_enable() and call from arch_float_enable() same as arch_float_disable()
  - kernel: add "options" argument into k_float_enable()

File: arch/x86/core/ia32/float.c [hunk 5]
```diff
@@ -179,7 +179,7 @@ static inline void FpCtxInit(struct k_thread *thread)
  * The locking isn't really needed when the routine is called by a cooperative
  * thread (since context switching can't occur), but it is harmless.
  */
-void k_float_enable(struct k_thread *thread, unsigned int options)
+void z_float_enable(struct k_thread *thread, unsigned int options)
```
Comment (dcpleung) [inline arch/x86/core/ia32/float.c:182]: Renaming it to `arch_float_enable()` will get rid of the `z_float_enable()` redirection.

Comment (katsuster) [reply arch/x86/core/ia32/float.c:182]: Thanks for your review. Yes, right. This is to keep consistency with existed k_float_disable() implementation.
If change such k_float_*() redirection of x86, it's better to change in an other commit (or PR ?).

Comment (dcpleung) [reply arch/x86/core/ia32/float.c:182]: It's fine to update it in another PR.

**Review by dcpleung — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
@katsuster pls, take a look at the compliance check failures

**Comment by katsuster** [comment]
@ioannisg 
Thanks for comments. It seems scripts/checkpatch.pl always warned if we use ENOSYS in the C code.
I think we can ignore this compliance check error at this time.

**Review by ghost — APPROVED** [review]

**Comment by pabigot** [comment]
Regarding `-ENOSYS` #29806 and particularly #23727 may be relevant.

I have no position on this PR nor on whether the checkpatch diagnostics on `-ENOSYS` should be removed.  Zephyr has been moving (if not entirely consistently) away from blindly following Linux coding standards, so any solution should be considered.

**Comment by pabigot** [comment]
Based on the information provided  in #23727 I believe the Zephyr standard solution would be to use `-ENOTSUP` instead of `-ENOSYS` for the return value, avoiding the checkpatch error which is useful as it identifies cases where the return value is not consistent with Zephyr standard.  (The diagnostic message could be improved, though.)

**Comment by carlescufi** [comment]
@katsuster we concluded that you should replace `ENOSYS` with `ENOTSUP`.
https://github.com/zephyrproject-rtos/zephyr/issues/23727#issuecomment-805051697

**Comment by katsuster** [comment]
@pabigot Thanks for describing about ENOSYS and ENOTSUP.
@carlescufi Thanks a lot. I understand. I will fix my patch.

**Comment by pabigot** [comment]
#33637 updates the checkpatch behavior.

**Comment by nashif** [comment]
@ruuddw  @evgeniy-paltsev @abrodkin  please review for arc

**Comment by ruuddw** [comment]
> @ruuddw @evgeniy-paltsev @abrodkin please review for arc

The PR said "The arc and riscv already implemented arch_float_enable() so I do not touch these implementations." so I didn't pay much attention, thanks for explicitly asking. I see an "option" parameter has been added (used by x86), that's fine.

**Review by ruuddw — APPROVED** [review]

**Pull request merged**